### PR TITLE
Abortable scale-down

### PR DESCRIPTION
**Pull request by otrosien**
# One-line summary

The ES Operator could decide to scale down, but while scaling down the thresholds for scale up are being exceeded, e.g. if we're reaching a point where we should stop scaling down.

## References

Issue : #57 

## Description

Introduces StatefulResourceGetter interface to allow refreshing of an EDS during scaling, which happens in case the autoscaler indicated a scale-up.

## Types of Changes

- New feature (non-breaking change which adds functionality)
- Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: operator/operator.go [hunk 1]
```diff
@@ -457,6 +479,17 @@ func (o *Operator) rescaleStatefulSet(ctx context.Context, sts *appsv1.StatefulS
...
+			newSR, err := srg.Get(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to refresh EDS: %v", err)
+			}
+			newDesiredReplicas := int(newSR.Replicas())
+			if newDesiredReplicas > desiredReplicas {
+				log.Infof("EDS %s/%s target scaling definition changed from %d to %d, aborting scale-down", sr.Namespace(), sr.Name(), desiredReplicas, newDesiredReplicas)
+				return nil
+			}
+
```
Comment (mikkeloscar) [inline operator/operator.go:492]: nit: I would propose to move this to the beginning of the loop just to make it more clear that we check this before scaling down anything.

Comment (otrosien) [reply operator/operator.go]: done.

**Comment by otrosien** [comment]
@mikkeloscar what else would you like to have considered in this PR?

**Comment by otrosien** [comment]
👍

File: operator/operator.go [hunk 2]
```diff
@@ -438,7 +459,19 @@ func (o *Operator) rescaleStatefulSet(ctx context.Context, sts *appsv1.StatefulS
...
+		log.Infof("Starting pod draining from %d to %d pods", len(pods.Items), replicas)
 		for _, pod := range pods.Items[replicas:] {
+			// first, check if we need to opt-out of the loop because the EDS changed.
+			newSR, err := srg.Get(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to refresh EDS: %v", err)
+			}
+			newDesiredReplicas := int(newSR.Replicas())
+			if newDesiredReplicas > desiredReplicas {
+				log.Infof("EDS %s/%s target scaling definition changed from %d to %d, aborting scale-down", sr.Namespace(), sr.Name(), desiredReplicas, newDesiredReplicas)
```
Comment (mikkeloscar) [inline operator/operator.go:471]: probably more clear if we use `newSR` here instead of `sr`. Same when calling `Drain` on line 488 below.

Comment (otrosien) [reply operator/operator.go]: fixed

**Comment by otrosien** [comment]
👍

**Comment by mikkeloscar** [comment]
:+1:

**Pull request merged**